doc/maximum_weighted_matching.html

```diff
@@ -0,0 +1,163 @@
+<html><head><!--
+     Copyright 2018 Yi Ji
+    
+     Use, modification and distribution is subject to the Boost Software
+     License, Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
+     http://www.boost.org/LICENSE_1_0.txt)
+    
+      Author: Yi Ji
+  --><title>Boost Graph Library: Maximum Weighted Matching</title></head>
+<body alink="#ff0000" bgcolor="#ffffff" link="#0000ee" text="#000000" vlink="#551a8b">
+<img src="../../../boost.png" alt="C++ Boost" height="86" width="277">
+<br clear="">
+<h1>
+<a name="sec:maximum_weighted_matching">Maximum Weighted Matching</a>
+</h1>
+<pre>
+template &lt;typename Graph, typename MateMap&gt;
+void maximum_weighted_matching(const Graph&amp; g, MateMap mate);
+
+template &lt;typename Graph, typename MateMap, typename VertexIndexMap&gt;
+void maximum_weighted_matching(const Graph&amp; g, MateMap mate, VertexIndexMap vm);
+
+template &lt;typename Graph, typename MateMap&gt;
+void brute_force_maximum_weighted_matching(const Graph&amp; g, MateMap mate);
+
+template &lt;typename Graph, typename MateMap, typename VertexIndexMap&gt;
+void brute_force_maximum_weighted_matching(const Graph&amp; g, MateMap mate, VertexIndexMap vm);
+</pre>
+<p>
+<a name="sec:weighted_matching">Before you continue, it is recommended to read
+about <a href="./maximum_matching.html">maximal cardinality matching</a> first. 
+A <i>maximum weighted matching</i> of an edge-weighted graph is a matching
+for which the sum of the weights of the edges is maximum. 
+Two different matchings (edges in the matching are colored blue) in the same graph are illustrated below. 
+The matching on the left is a maximum cardinality matching of size 8 and a maximal
+weighted matching of weight sum 30, meaning that is has maximum size over all matchings in the graph
+and its weight sum can't be increased by adding edges.
+The matching on the right is a maximum weighted matching of size 7 and weight sum 38, meaning that it has maximum
+weight sum over all matchings in the graph.
```

Comment:
In terms of nomenclature, Gabow (1990) mentions a _maximum weight matching_ problem but in his earlier 1976 paper the problem is called _maximum matching_, as it is by Edmonds in his original paper.
My guess is that he added "weight" in to differentiate against the _maximum cardinality matching_ problem.
But either way, I'm going to suggest that it is "weight", not "weighted", although I know he uses the latter in some other names. Anyway, it's not a serious issue, just my pedantic opinion.

In addition to that, the 1990 paper raises another point, that this O(n^3) algorithm is not the only algorithm that solves maximum weight matching, as it presents an O(n(m + n log n)) algorithm. It's perfectly reasonable to implement the O(n^3) algorithm, but we need to name it such that we can also name the newer algorithm, e.g.:

- `brute_force_maximum_weight_matching` -- O(?)
- `label_encoding_maximum_weight_matching` -- O(n^3)
- `blossom_expansion_maximum_weight_matching` -- O(n(m + n log n))

They are just ideas, I'm not sure if they are entirely accurate, especially the last one.

Then, maybe, we could have a `maximum_weight_matching` that uses some heuristics to pick the fastest algorithm for a given problem size. Just an idea, not important.

Reply:
I believe both "maximum weight / weighted matching" are commonly used. In fact, another large C++ template library for network algos, [LEMON](http://lemon.cs.elte.hu/pub/doc/latest/a00254.html), uses `MaxWeightedMatching`.

Thanks for the name prefix suggestion. However the examples might not be quite accurate. The O(n^3) and O(n(m + n log n)) algorithms both have labeling and blossom expansion operations. The only difference is the O(n(m + n log n)) algorithms uses link-cut-tree to find nearest common ansestors, which is not important/radical/obvious enough to call them with different names.
I didn't implement the O(n(m + n log n)) algo mainly because there is no available link-cut-tree module in boost. If it becomes available and someone wants to improve to O(n(m + n log n)) someday, he/she could just make a few lines of code change and doesn't have to keep an O(n^3) version.

Reply:
Ah, hmmm, OK, you think the difference between the O(n^3) and O(n(m + n log n)) algorithms will be a matter of passing in different LCA sub-algorithms? That would certainly make the naming easier.

Funnily enough, I am actually working on... close to finishing... a pragmatic implementation of LCA, but it's not Gabow's  link-cut version.  When I have a chance, I'll have a look at how much more work it would be to extend it, although I want to finish the basic implementation first.

Reply:
Thanks. More funnily... I am also working on and close to finishing a link-cut-tree module😂 but I believe it’s okay if we open dIfferent PR. Let’s see what to do then.